src/neptune/core/operations/operation.py

```diff
@@ -0,0 +1,212 @@
+#
+# Copyright (c) 2024, Neptune Labs Sp. z o.o.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+__all__ = [
+    "AssignInt",
+    "AssignFloat",
+    "AssignBool",
+    "AssignString",
+    "AssignDatetime",
+    "LogFloats",
+    "Operation",
+]
+
+import abc
+from dataclasses import dataclass
+from datetime import datetime
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    ClassVar,
+    Dict,
+    Generic,
+    List,
+    Optional,
+    Type,
+    TypeVar,
+    Union,
+)
+
+from neptune.core.components.operation_storage import OperationStorage
+from neptune.exceptions import MalformedOperation
+
+if TYPE_CHECKING:
+    from neptune.core.operations.operation_visitor import OperationVisitor
+
+Ret = TypeVar("Ret")
+T = TypeVar("T")
+
+
+@dataclass
+class Operation(abc.ABC):
+    path: List[str]
+    _registry: ClassVar[Dict[str, Type["Operation"]]] = {}
+
+    def __init_subclass__(cls, **kwargs: Any) -> None:
+        super().__init_subclass__(**kwargs)
+        cls._registry[cls.__name__] = cls
+
+    @abc.abstractmethod
+    def accept(self, visitor: "OperationVisitor[Ret]") -> Ret:
+        pass
+
+    def clean(self, operation_storage: OperationStorage) -> None:
+        pass
+
+    def to_dict(self) -> dict:
+        return {"type": self.__class__.__name__, "path": self.path}
+
+    @staticmethod
+    def from_dict(data: dict) -> "Operation":
+        if "type" not in data:
+            raise MalformedOperation("Malformed operation {} - type is missing".format(data))
+        operation_type = data["type"]
+        if operation_type not in Operation._registry:
+            raise MalformedOperation("Malformed operation {} - unknown type {}".format(data, operation_type))
+        return Operation._registry[operation_type].from_dict(data)
+
+
+@dataclass
+class AssignFloat(Operation):
+    value: float
+
+    def accept(self, visitor: "OperationVisitor[Ret]") -> Ret:
+        return visitor.visit_assign_float(self)
+
+    def to_dict(self) -> dict:
+        ret = super().to_dict()
+        ret["value"] = self.value
+        return ret
+
+    @staticmethod
+    def from_dict(data: dict) -> "AssignFloat":
+        return AssignFloat(data["path"], data["value"])
+
+
+@dataclass
+class AssignInt(Operation):
+
+    value: int
```

Comment:
some classes have extra space and some do not. Did you run formatter on all those files?

Reply:
I think we good rn